http-server/src/main/java/io/micronaut/http/server/OptionsFilter.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.server;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.core.order.Ordered;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.*;
+import io.micronaut.http.annotation.RequestFilter;
+import io.micronaut.http.annotation.ServerFilter;
+import io.micronaut.http.server.cors.CorsUtil;
+import io.micronaut.web.router.Router;
+import io.micronaut.web.router.UriRouteMatch;
+import io.micronaut.web.router.RouteMatch;
+
+import static io.micronaut.http.annotation.Filter.MATCH_ALL_PATTERN;
+import static io.micronaut.http.server.cors.CorsFilter.CORS_FILTER_ORDER;
+
+/**
+ * This Filter intercepts HTTP OPTIONS requests which are not CORS Preflight requests.
+ * It responds with a NO_CONTENT(204) response, and it populates the Allow HTTP Header with the supported HTTP methods for the request URI.
+ * @author Sergio del Amo
+ * @since 4.2.0
+ */
+@Requires(property = OptionsFilter.PREFIX, value = StringUtils.TRUE, defaultValue = StringUtils.FALSE)
+@ServerFilter(MATCH_ALL_PATTERN)
+@Internal
+public final class OptionsFilter implements Ordered {
+
+    @SuppressWarnings("WeakerAccess")
+    public static final String PREFIX = HttpServerConfiguration.PREFIX + ".dispatch-options-requests";
+
+    private final Router router;
+
+    /**
+     *
+     * @param router Router
+     */
+    public OptionsFilter(Router router) {
+        this.router = router;
+    }
+
+    @RequestFilter
+    @Nullable
+    @Internal
+    public HttpResponse<?> filterRequest(HttpRequest<?> request) {
+        if (request.getMethod() != HttpMethod.OPTIONS) {
+            return null; // proceed
+        }
+        if (CorsUtil.isPreflightRequest(request)) {
+            return null; // proceed
+        }
+        if (hasOptionsRouteMatch(request)) {
+            return null; // proceed
+        }
+        MutableHttpResponse<?> mutableHttpResponse = HttpResponse.status(HttpStatus.NO_CONTENT);
```

Comment:
While I agree that a `204` response code seems at first to be more semantically correct, I think it is more widely compatible and safer to return a `200` with `Content-Length` of 0. 

The HTTP spec does not explicitly state the status code to use for a general successful OPTIONS request, but web browsers have historically exhibited some strange behavior when handling a `204` in response to CORS pre-flight OPTIONS requests, and the general HTTP 1.1 semantics spec (RFC 7231) *used* to say that a `Content-Length` header of 0 must be set on a successful  OPTIONS response, and the `Content-Length` header is not allowed on a `204` response.  Thus I think there are a number of existing clients that somewhat expect the `200`. 

The most recent relevant spec (https://www.rfc-editor.org/rfc/rfc9110) explicitly removed the `Content-Length` header requirement, but the description of `200` makes explicit mention of OPTIONS requests, whereas the `204` description does not.

Reply:
Please note this filter does not handle CORS preflight requests. This PR is unrelated to CORS.

Reply:
@sdelamo Yes, I understand that. I was using CORS as a specific example of known strange behavior of clients, but I can see how it confused matters.

The HTTP spec definitions I was referring to have nothing to do with CORS either. 

Even the most recently updated version of the specs are (and have historically been) ambiguous on this point, but I think the superfluous inclusion of the `Content-Length` header requirement in [RFC 7231](https://www.rfc-editor.org/rfc/rfc7231#section-4.3.7) led some clients to implement OPTIONS requests in a way that expects a 200 response.

Re-reading the most recent version ([RFC 9110 - 15.3.5](https://www.rfc-editor.org/rfc/rfc9110#status.204)) again, I notice the wording of the spec as regards the 204 status is really targeted more towards requests that result in some kind of actual action on the resource in question, such as a PUT or DELETE.

Whereas the description of the [200 status code](https://www.rfc-editor.org/rfc/rfc9110#status.200) explicitly states that a 200 response to indicates that the "intended meaning of the content" in response to an OPTIONS request is "communication options for the target resource" which is very much what is being supplied by this filter.

Reply:
[I have changed to use 200 instead of 204](https://github.com/micronaut-projects/micronaut-core/pull/10011/commits/c05064292883cf6a7b928ddcd65ebe17cb4479e0)